Added option to disable signal handler

DavidPfander-UniStuttgart opened the pull request:
Allows errors to get to the user unhandled, applies to exceptions and
signals. For some people, this makes it easier to debug segfaults and
unhandled exception. With this commit the default behavior isn't
changed, this behavior has to be explictly enabled through a cmake option.

hkaiser commented on CMakeLists.txt:
```diff
@@ -333,6 +333,15 @@ if(HPX_WITH_AUTOMATIC_SERIALIZATION_REGISTRATION)
+
+if(NOT HPX_WITH_SIGNAL_HANDLERS)
+  add_definitions(-DHPX_DISABLE_SIGNAL_HANDLERS)
+endif()
```
The convention in HPX is to name the PP constant corresponding to a `HPX_WITH_FOO` configuration setting as `HPX_HAVE_FOO`. Could you change this name accordingly, please? I'd suggest to name it `HPX_HAVE_NO_SIGNAL_HANDLERS` or somesuch. Please also align the sames such that the configuration option is named similarily (i.e. `HPX_WITH_NO_SIGNAL_HANDLERS`).

DavidPfander-UniStuttgart commented:
I made the names of the PP directive and the cmake option consistent (== the same) and retested (also fixed a type and a small bug).

sithhell commented:
Could you also add documentation for that?

Am 03.02.2017 5:07 nachm. schrieb "DavidPfander-UniStuttgart" <
notifications@github.com>:

> I made the names of the PP directive and the cmake option consistent (==
> the same) and retested (also fixed a type and a small bug).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/STEllAR-GROUP/hpx/pull/2479#issuecomment-277288050>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADRl6AdgksM10xpHoNm52DcVbms-RbCks5rY1DJgaJpZM4L2aEt>
> .
>

DavidPfander-UniStuttgart commented:
Added documentation modeled on the option right before the new option. Unfortunatel, as I don't have the documentation toolchain installed, I wasn't able to test it.

hkaiser commented on CMakeLists.txt:
```diff
@@ -333,6 +333,15 @@ if(HPX_WITH_AUTOMATIC_SERIALIZATION_REGISTRATION)
 
+hpx_option(HPX_WITH_DISABLED_SIGNAL_EXCEPTION_HANDLERS BOOL
+  "Disables the mechanism that produces debug output for catched signals and unhandled exceptions (default: OFF)"
+  OFF
```
The past tense of 'catch' is 'caught'.

hkaiser commented on docs/manual/build_system/cmake_variables.qbk:
```diff
@@ -66,6 +67,7 @@ The options are split into these categories:
         [[[#build_system.cmake_variables.HPX_WITH_ASYNC_FUNCTION_COMPATIBILITY] `HPX_WITH_ASYNC_FUNCTION_COMPATIBILITY:BOOL`][Enable old style ..._sync/..._async functions in API (default: OFF)]]
         [[[#build_system.cmake_variables.HPX_WITH_AUTOMATIC_SERIALIZATION_REGISTRATION] `HPX_WITH_AUTOMATIC_SERIALIZATION_REGISTRATION:BOOL`][Use automatic serialization registration for actions and functions. This affects compatibility between HPX applications compiled with different compilers (default ON)]]
+        [[[#build_system.cmake_variables.HPX_WITH_DISABLED_SIGNAL_EXCEPTION_HANDLERS] `HPX_WITH_DISABLED_SIGNAL_EXCEPTION_HANDLERS:BOOL`][Disables the mechanism that produces debug output for signals and unhandled exceptions (default: OFF)]]
         [[[#build_system.cmake_variables.HPX_WITH_BENCHMARK_SCRIPTS_PATH] `HPX_WITH_BENCHMARK_SCRIPTS_PATH:PATH`][Directory to place batch scripts in]]
```
This line should get automatically generated, so it will have the same descriptive text as specified in the cmake file. No need to manually edit those.

hkaiser commented:
Also: inspect is not happy with you adding tabs, see: https://5896-4455628-gh.circle-artifacts.com/0/tmp/circle-artifacts.EmGfXMK/hpx_inspect_report.html

sithhell commented:
What I meant with documenting the option was to add a section to the manual about debugging HPX that explicitly explains the implications of this option and when users want to set set etc. This could be the starting point for other debugging related cmake options and command line parameters.

DavidPfander-UniStuttgart commented:
Reverted my documentation changes and fixed the type Hartmut mentioned. 

@sithhell:
I'll happily add a paragraph about this option. Where in the documentation should I add it?

hkaiser commented:
@DavidPfander-UniStuttgart perhaps this could be added here: http://stellar-group.github.io/hpx/docs/html/hpx/manual/build_system/building_hpx.html (i.e. here: https://github.com/STEllAR-GROUP/hpx/blob/master/docs/manual/build_system.qbk)

hkaiser commented:
Ok, fine by me. So the last things would be to make inspect happy: https://6025-4455628-gh.circle-artifacts.com/0/tmp/circle-artifacts.cY8L1hs/hpx_inspect_report.html.

sithhell commented:
On 02/17/2017 01:11 PM, Hartmut Kaiser wrote:
> Ok, fine by me. So the last things would be to make inspect happy:
> https://6025-4455628-gh.circle-artifacts.com/0/tmp/circle-artifacts.cY8L1hs/hpx_inspect_report.html.

That's due to my messup on master ... #2506 contains a patch for this.
My appologies.
